db/migrate/20250912000001_populate_pulp_prn_fields.rb

```diff
@@ -0,0 +1,377 @@
+          'prn:ansible.ansiblerepository:' ||
+          regexp_replace(repository_href, '^.*/([^/]+)/?$', '\\1')
+        WHEN repository_href ~ '/repositories/python/python/' THEN
+          'prn:python.pythonrepository:' ||
+          regexp_replace(repository_href, '^.*/([^/]+)/?$', '\\1')
+        WHEN repository_href ~ '/repositories/file/file/' THEN
+          'prn:file.filerepository:' ||
+          regexp_replace(repository_href, '^.*/([^/]+)/?$', '\\1')
+        ELSE NULL
+      END"
+    )
+
+    # Repository Errata - erratum_prn
+    Katello::RepositoryErratum.where.not(erratum_pulp3_href: nil).update_all(
+      "erratum_prn = CASE
+        WHEN erratum_pulp3_href ~ '/content/rpm/advisories/' THEN
+          'prn:rpm.updaterecord:' ||
+          regexp_replace(erratum_pulp3_href, '^.*/([^/]+)/?$', '\\1')
+        ELSE NULL
+      END"
+    )
+
+    # Smart Proxy Alternate Content Sources - remote_prn
+    Katello::SmartProxyAlternateContentSource.where.not(remote_href: nil).update_all(
+      "remote_prn = CASE
+        WHEN remote_href ~ '/remotes/rpm/rpm/' THEN
+          'prn:rpm.rpmremote:' ||
+          regexp_replace(remote_href, '^.*/([^/]+)/?$', '\\1')
+        WHEN remote_href ~ '/remotes/rpm/uln/' THEN
+          'prn:rpm.ulnremote:' ||
+          regexp_replace(remote_href, '^.*/([^/]+)/?$', '\\1')
+        WHEN remote_href ~ '/remotes/deb/apt/' THEN
+          'prn:deb.aptremote:' ||
+          regexp_replace(remote_href, '^.*/([^/]+)/?$', '\\1')
+        WHEN remote_href ~ '/remotes/container/container/' THEN
+          'prn:container.containerremote:' ||
+          regexp_replace(remote_href, '^.*/([^/]+)/?$', '\\1')
+        WHEN remote_href ~ '/remotes/ansible/collection/' THEN
+          'prn:ansible.collectionremote:' ||
+          regexp_replace(remote_href, '^.*/([^/]+)/?$', '\\1')
+        WHEN remote_href ~ '/remotes/python/python/' THEN
+          'prn:python.pythonremote:' ||
+          regexp_replace(remote_href, '^.*/([^/]+)/?$', '\\1')
+        WHEN remote_href ~ '/remotes/file/file/' THEN
+          'prn:file.fileremote:' ||
+          regexp_replace(remote_href, '^.*/([^/]+)/?$', '\\1')
+        ELSE NULL
+      END"
+    )
+
+    # Smart Proxy Alternate Content Sources - alternate_content_source_prn
+    Katello::SmartProxyAlternateContentSource.where.not(alternate_content_source_href: nil).update_all(
+      "alternate_content_source_prn = CASE
+        WHEN alternate_content_source_href ~ '/acs/rpm/rpm/' THEN
+          'prn:rpm.rpmalternatecontentsource:' ||
+          regexp_replace(alternate_content_source_href, '^.*/([^/]+)/?$', '\\1')
+        WHEN alternate_content_source_href ~ '/acs/file/file/' THEN
+          'prn:file.filealternatecontentsource:' ||
+          regexp_replace(alternate_content_source_href, '^.*/([^/]+)/?$', '\\1')
+        ELSE NULL
+      END"
+    )
+
+    # # Add NOT NULL constraints to PRN fields that correspond to required href fields
```

Comment:
I think we can actually enable the repository version constraints at this point in the migration as well.

repository.version_href is nullable and has no unique constraint, so we could just leave the fields empty.

That way, the entire upgrade (besides removing the old href fields) could happen in a single release. After the migration, the version_href fields will be empty, and then a rake script triggered by the installer can populate them.

Reply:
Edit - I realized we will break dev environments if we include the constraints before the PRN indexing code is introduced. So, these can be introduced in that PR if it comes later.